Comment on pull requests

samcoe opened the pull request:
This PR allows for commenting on pull requests. The commenting code for issues and pull requests was very similar so I refactored to extract the `commentable.go` file which contains the shared functionality.

Additionally after talking with @ampinsk we decided that the initial input type survey was not serving much of a purpose and we decided to replace it with the standard editor survey that prompts the user to open their editor.

closes https://github.com/cli/cli/issues/517

samcoe committed:
Comment on pull requests

mislav reviewed (commented):
Is there a chance we could reuse the same code, since everything is identical about leaving a comment to an issue vs. the PR?

I worry that this might not be just incidental duplication: we might need to maintain the commenting code, and it might be tricky to remember to duplicate all changes in two places, especially for outside contributors.

samcoe commented:
@mislav I went ahead and refactored this to clean up the duplicate code and hopefully make it easier to maintain both comment commands in the future. It seemed like there were a couple different ways to go about this refactor so let me know if you had any other thoughts on it 🙇

mislav commented on pkg/cmd/pr/comment/comment.go:
```diff
@@ -0,0 +1,78 @@
+	return cmd
+}
+
+func retrievePR(httpClient func() (*http.Client, error),
+	baseRepo func() (ghrepo.Interface, error),
+	branch func() (string, error),
+	remotes func() (context.Remotes, error),
+	selector string) func() (shared.Commentable, ghrepo.Interface, error) {
+	return func() (shared.Commentable, ghrepo.Interface, error) {
```
I feel that this function's signature is extra complex, but I also understand that in order to unwrap it, we would have to forfeit the lazy-loading aspect of the whole thing. So, refactoring this into something cleaner isn't something that I know how to suggest from the top of my head, but we should maybe keep in mind to try to do it as followup, maybe via pairing? Anyway, if it feels stable to you, then I'm fine with shipping 👍

samcoe replied:
I agree, that it is overly complex in its current form and is a good area for refactoring later on. I am confident that it is stable and okay to ship 👍

mislav commented on pkg/cmd/pr/shared/commentable.go:
```diff
@@ -0,0 +1,163 @@
+			return err
+		}
+		if !cont {
+			return fmt.Errorf("Discarding...")
```
Nit: if the error doesn't need to be formatted, you can use `errors.New()`.

mislav commented on pkg/cmd/pr/shared/commentable.go:
```diff
@@ -0,0 +1,163 @@
+				FileName: "*.md",
+			},
+		}
+		err = survey.AskOne(prompt, &body)
```
I know that I might sound repetitive by re-raising the same issue as before, but I must admit that I still do not understand the purpose of this prompt. I'm being asked this:

```
$ gh pr comment 123
? Body [(e) to launch vim]
```

My options are: type <kbd>e</kbd> to launch my text editor, or press <kbd>Ctrl-C</kbd> to bail out. Now, I'm bringing this up because I'm not sure what value this prompt brings to the commenting flow? When this prompt appears during `gh issue/pr create`, it makes sense to ask to open the editor because the user can very well choose to skip it since it's optional. Skipping writing a body makes no sense for an action (commenting) whose only input is a written body.

Couldn't we just open an editor? If the user wants to bail out at that point, they can just close their editor.

I'm worried that being liberal with adding prompts around GitHub CLI will lead to prompt fatigue for our users down the line.

ampinsk replied:
I felt the prompt was needed here since jumping into the editor straight from a command being run would be jarring. This doesn't necessarily need to be a `survey` prompt, I just wanted to give users an opportunity to knowingly jump into their editor. Not sure what technically would make sense here but it could simply be:

```
$ gh pr comment 123
Press **enter** to draft your comment in {editor}
```

mislav replied:
Thank you for explaining. If the purpose is to provide a keypress guard before opening a text editor program, then the prompt would make sense, and I guess it wouldn't matter much whether the specific key was <kbd>e</kbd> or <kbd>Enter</kbd>.

With my comment, I'm just amplifying the perspective of users who would prefer if the editor was launched right away at times when we know that the user will want to write multi-line text: https://github.com/cli/cli/issues/390 https://github.com/cli/cli/issues/1090. I personally would also prefer that we skip a prompt. But of course, I'll leave the decision up to you.

mislav commented on pkg/cmd/pr/comment/comment.go:
```diff
@@ -0,0 +1,78 @@
+		Example: heredoc.Doc(`
+			$ gh pr comment 22 --body "This looks great, lets get it deployed."
+		`),
+		PreRunE: func(cmd *cobra.Command, args []string) error {
```
I've never figured out a good use-case for Cobra's hooks. I like that you're exploring this here. But to help me understand: is this PreRunE block here because the same logic wouldn't work (or wouldn't be as testable) if it were just moved to the RunE block?

samcoe replied:
I used the `PreRunE` block here mostly for organization and readability of the code. The same code could be moved into the `RunE` block and everything would function the exact same. I felt that in the `Commentable` file we needed a clean way to differentiate between the configuring of the command and the running of the command, this distinction is easier to follow when initializing the `cobra.Command` but got lost when extracting the common code to `Commentable`. Overall, this direction felt cleaner and created an easy mapping between `Commentable`  functions and the PR comment/issue comment functions.

mislav reviewed (changes requested):
This looks great! Love it how you reused most commenting code between issues and PRs by defining a Commentable interface.

My last nit is about the prompt to open the editor. Everything else is 💎💎

mislav committed:
Bring the "Press Enter" UI closer to the authentication experience

- "Press Enter" is both bold
- "Enter" is capitalized
- The prompt ends with "..."

mislav reviewed (approved):
Updates look good. Thank you for the hard work!

Louisianaanimal commented:
Hello crypto fam
